### PR TITLE
Add relevant checks for tuple comparison.

### LINKED ISSUE #53797: Address the PROTOTYPE comment in BindTupleBinaryOperatorInfo
```
        private TupleBinaryOperatorInfo BindTupleBinaryOperatorInfo(BinaryExpressionSyntax node, BinaryOperatorKind kind,
            BoundExpression left, BoundExpression right, BindingDiagnosticBag diagnostics)
        {
            TypeSymbol leftType = left.Type;
            TypeSymbol rightType = right.Type;

            if ((object)leftType != null && leftType.IsDynamic() || (object)rightType != null && rightType.IsDynamic())
            {
                return BindTupleDynamicBinaryOperatorSingleInfo(node, kind, left, right, diagnostics);
            }

            if (IsTupleBinaryOperation(left, right))
            {
                return BindTupleBinaryOperatorNestedInfo(node, kind, left, right, diagnostics);
            }

            BoundExpression comparison = BindSimpleBinaryOperator(node, diagnostics, left, right);
            switch (comparison)
            {
                case BoundLiteral _:
                    // this case handles `null == null` and the like
                    return new TupleBinaryOperatorInfo.NullNull(kind);

                case BoundBinaryOperator binary:
                    PrepareBoolConversionAndTruthOperator(binary.Type, node, kind, diagnostics, out Conversion conversionIntoBoolOperator, out UnaryOperatorSignature boolOperator);
                    // PROTOTYPE(StaticAbstractMembersInInterfaces): Ensure we have a unit-test for this code path.
                    return new TupleBinaryOperatorInfo.Single(binary.Left.Type, binary.Right.Type, binary.OperatorKind, binary.MethodOpt, binary.ConstrainedToTypeOpt, conversionIntoBoolOperator, boolOperator);

                default:
                    throw ExceptionUtilities.UnexpectedValue(comparison);
            }
        }
```

### LINKED ISSUE #53799: Address the PROTOTYPE comments in LocalRewriter.RewriteTupleSingleOperator
```
            // PROTOTYPE(StaticAbstractMembersInInterfaces): MAke sure we have test coverage for this code path
            BoundExpression binary = MakeBinaryOperator(_factory.Syntax, single.Kind, left, right, single.MethodSymbolOpt?.ReturnType ?? boolType, single.MethodSymbolOpt, single.ConstrainedToTypeOpt);
            UnaryOperatorSignature boolOperator = single.BoolOperator;
            Conversion boolConversion = single.ConversionForBool;

            BoundExpression result;
            if (boolOperator.Kind != UnaryOperatorKind.Error)
            {
                // Produce
                // !((left == right).op_false)
                // (left != right).op_true
                BoundExpression convertedBinary = MakeConversionNode(_factory.Syntax, binary, boolConversion, boolOperator.OperandType, @checked: false);

                Debug.Assert(boolOperator.ReturnType.SpecialType == SpecialType.System_Boolean);
                // PROTOTYPE(StaticAbstractMembersInInterfaces): MAke sure we have test coverage for this code path
                result = MakeUnaryOperator(boolOperator.Kind, binary.Syntax, boolOperator.Method, boolOperator.ConstrainedToTypeOpt, convertedBinary, boolType);
```

### PR DESCRIPTION
**Pull request by AlekseyTs**
Closes #53797.
Closes #53799.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@333fred, @RikkiGibson, @dotnet/roslyn-compiler Please review.

**Comment by AlekseyTs** [comment]
@333fred, @RikkiGibson, @dotnet/roslyn-compiler Please review.

**Comment by AlekseyTs** [comment]
@333fred, @RikkiGibson, @dotnet/roslyn-compiler Please review.

**Comment by AlekseyTs** [comment]
@333fred, @RikkiGibson, @dotnet/roslyn-compiler Please review.

File: src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs [hunk 1]
```diff
@@ -9191,6 +9627,183 @@ .maxstack  2
...
+@"
+public interface I1<T> where T : I1<T>
+{
+    abstract static bool operator == (T x, T y);
+    abstract static bool operator != (T x, T y);
+}
+
+class Test
+{
+    static void M02<T, U>((int, T) x) where T : U where U : I1<T>
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs:9644]: I was surprised that this worked. I thought `T` needed to be directly constrained to `I1<T>` to access abstract static members on it for the time being.

File: src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs [hunk 2]
```diff
@@ -447,6 +447,7 @@ private BoundForEachStatement BindForEachPartsWorker(BindingDiagnosticBag diagno
             else
             {
                 ReportDiagnosticsIfObsolete(diagnostics, elementConversion, _syntax.ForEachKeyword, hasBaseReceiver: false);
+                CheckConstraintLanguageVersionAndRuntimeSupportForConversion(_syntax.ForEachKeyword, elementConversion, diagnostics);
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:450]: I didn't see tests in `StaticAbstractMembersInInterfacesTests.cs` for this code path. It seems like the test would be a foreach loop which subjects the collection element to an abstract static conversion, and either the language version or runtime doesn't support the usage of the conversion.

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs [hunk 3]
```diff
@@ -3844,6 +3848,7 @@ private BoundExpression BindNullCoalescingOperator(BinaryExpressionSyntax node,
                     else
                     {
                         ReportDiagnosticsIfObsolete(diagnostics, leftConversion, node, hasBaseReceiver: false);
+                        CheckConstraintLanguageVersionAndRuntimeSupportForConversion(node, leftConversion, diagnostics);
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:3851]: I didn't see usage of null coalescing operators in `StaticAbstractMembersInInterfacesTests.cs`. Similar to the foreach element scenario, should we have a test which ensures we report appropriate LangVersion and runtime support errors on these null coalescing operand conversions?

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs [hunk 4]
```diff
@@ -214,6 +215,25 @@ internal partial class Binder
             { WasCompilerGenerated = wasCompilerGenerated };
         }
 
+        internal void CheckConstraintLanguageVersionAndRuntimeSupportForConversion(SyntaxNodeOrToken syntax, Conversion conversion, BindingDiagnosticBag diagnostics)
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:218]: I am a little curious about why this check had to move, as well as how you determined where are all the appropriate places to insert a call to this method. Are there simply other places where user-defined conversions are created that we had to account for?

File: src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs:9644]: > I was surprised that this worked. I thought T needed to be directly constrained to I1<T> to access abstract static members on it for the time being.

This isn't an operator, whih have signature constraints, and   it is not declared in an interface.

File: src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:450]: > I didn't see tests in StaticAbstractMembersInInterfacesTests.cs for this code path. It seems like the test would be a foreach loop which subjects the collection element to an abstract static conversion, and either the language version or runtime doesn't support the usage of the conversion.

Testing this is tracked by [https://github.com/dotnet/roslyn/issues/53870](https://github.com/dotnet/roslyn/issues/53870). Ther are a lot of similar scenarios to cover, I am not planning to do this in this PR, but I am pretty comfortable to add obvious checks that are working for other tested scenarios.

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs (hunk 3 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:3851]: > I didn't see usage of null coalescing operators in StaticAbstractMembersInInterfacesTests.cs. Similar to the foreach element scenario, should we have a test which ensures we report appropriate LangVersion and runtime support errors on these null coalescing operand conversions?

Tracked by the same issue as the other.

**Review by 333fred — APPROVED** [review]
LGTM (commit 1), since added tests are tracked.

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs (hunk 4 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:218]: > I am a little curious about why this check had to move, as well as how you determined where are all the appropriate places to insert a call to this method. Are there simply other places where user-defined conversions are created that we had to account for?

The original check was covering only BoundConversion node. However, there are a lot of other nodes that classify conversion, store Conversion instance and emulate precence of BoundConversion in lowering. That is how I located some of the places (theer was no goal to locate them all in this PR). Also, it felt appropriate to keep the checks close to the code that reports obsolete diagnostics for conversions (ReportDiagnosticsIfObsolete).

**Review by RikkiGibson — APPROVED** [review]

**Pull request merged**